backend/cfg/cfgize.ml

```diff
@@ -762,7 +762,7 @@ let fundecl :
      integer test. This simplification should happen *after* the one about
      straightline blocks because merging blocks creates more opportunities for
      terminator simplification. *)
-  Merge_straightline_blocks.run cfg_with_layout;
+  if simplify_terminators then Merge_straightline_blocks.run cfg_with_layout;
   Eliminate_dead_code.run_dead_block cfg_with_layout;
   if simplify_terminators then Simplify_terminator.run cfg;
```

Comment:
What is the rationale behind the order of the transformations in L765-767? Should `Eliminate_dead_code` run after the other two transformations?

Reply:
A transformation can sometimes create new
opportunities; are you suggesting to compute
a fix point?

Reply:
Actually, in this case, if the comment in the code
is to be trusted "merge" should happen before
"simplify", and I would expect "eliminate" to
create no new opportunity for the others. So, the
question might be: does "simplify" create new
opportunities for "merge"? I would say no, but I
am not positive.

Reply:
> A transformation can sometimes create new opportunities; are you suggesting to compute a fix point?

no, no, I don't want to have a fixpoint here, just trying to understand the code.

> Actually, in this case, if the comment in the code is to be trusted "merge" should happen before "simplify", 

I don't remember why the comment is true, but lets assume so. 

> and I would expect "eliminate" to create no new opportunity for the others. 

`Eliminate_dead_code` can create more opportunities for `Merge`. I don't think that "Eliminate"  creates more opportunities for `Simplify_terminator` directly.

> So, the question might be: does "simplify" create new opportunities for "merge"? I would say no, but I am not positive.

`Simplify_terminator` won't create new opportunities for `Merge_straightline_blocks` because `Merge_straightline_blocks` already looks at the number of (distinct) successors (which doesn't change by `Simplify_terminator`) and not at the "shape" of the terminator. 

I think the order should be:  eliminate, merge, simplify.

Reply:
Merge can absolutely create opportunities:
```
     [block ending with if/else] --> [empty block] --> [ ... ]
                               \                         ^
                                \-----------------------/
```

Reply:
I don't immediately see how "eliminate" can
create opportunities for "merge"; would you
have an example?

Reply:
[A]--> [B] --->[...]
[D]-----^

If D is dead, after eliminating it, A and B can be merged, but not before.

Reply:
In this particular case, we may be OK because
I doubt `Cfgize` will generate such graphs.

Reply:
Another perpective: if your change has an effect,
it means it will break the equivalence checks that
currently pass.

Reply:
The order of these transformations is the same in asmgen. We can deal with reordering of these transformations in a separate PR and consider refactoring all their occurrences to one place.